### DISCUSSION TITLE
"SVG-in-CSS": do we need a consistent policy regarding it?

### DISCUSSION THREAD

**Discussion by SelenIT** [cssbattle/forum#148] [General]
The current top solution for the latest Open Target ([#84](https://cssbattle.dev/play/84)) features a modern CSS `offset` property which accepts the syntax of the `d` attribute of the SVG `<path>` element as a value (in a string form). This provides a shorter replacement for the `transform: translate(...) rotate(...)` combination (and its equivalents) and saves some 10-15 characters of the solution. The UI of CSSBattle accepts such solutions perfectly fine.

AFAIK, there are 2 other CSS properties at the moment that also accept the SVG path data in a string form as a value and are supported by Chrome. One of them, `shape-outside`, is probably not very useful for CSS code golfing purposes (though I see some potential for it in targets like [#86](https://cssbattle.dev/play/86)), but the last one... well, it's behind many top solutions, and I'm also guilty in this. In some targets (esp. symmetric ones like "Batmicky"), it allowed to shorten the solution by some 30% relative to the second best approach I'm aware of. In previous discussions, we decided that such solutions should be disallowed, and the UI currently blocks the naive attempts to use this property and the most intuitive way to circumvent this restriction. However, another way to submit solutions with this property that requires only one extra character is still possible, and even if it's banned too, a workaround with storing the "forbidden" value in a custom property will always remain.

Given that it's still a _CSS_ syntax feature that is there in the specs [at least since 2014](https://drafts.csswg.org/css-shapes-1/#20140320) and has decent browser support, wouldn't it be better to allow it, but, e.g., limit the score for such solutions with 600 points – like for solutions of the same length without 100% match (to prevent overusing it for getting top scores)? And should the SVG syntax in different properties be treated the same way or not?..

**Comment by RazvanManolache** [comment]
Always read the documentation... offset has rotate too? Well, I would say it changes everything but I should learn how to use it. I saw it on the solution, played with the values, did a 'huh', went to bed. Now knowing it has rotate changes everything and nothing for me. I never heard of shape-outside either, I think I understand how it could use for Stairway, not that I could do it without investing too much effort into R&D.

I personally don't mind SVGs, from what I've been reading most solutions using SVG were beaten by good ol' CSS anyway.

Or they could make a separate ladder for that. You use any SVG syntax? BAM, score goes on the other ladder. Although that could be abused too, people adding some SVG code to their best plays just to add them to that one too. Unicode could be treated similarly too, but again, same exploit. Probably easier, just add a invisible unicode character and walk away.

Last option would be to add a cost, make it more interesting. Use SVG? How long? Deduct length times certain factor. For unicode the same. But then you have shadows. Yes, I only used a character but it has shadows all over. Probably will not have the same benefit tho.

There is no perfect way to do this, and I think it would spoil the game either way you do it. Because especially if you did all targets you know that top score requires knowledge but also persistence.

There would be another approach, to make it challenging. Something like open target but let's call it open tags. You get to see what tags/properties the person with the next bigger score used. Not the top one. The one with a slightly bigger score, even if it's by 0.01. Not the code, mind you. Or values. Or order. Or how many times they were used. Just tags/properties. That would actually provide a learning experience for people not knowing certain CSS features and it would still be a puzzle to figure out how to do it. You'd still be stuck in percentage/unit hell because someone figure out that em would work for a specific value or gradient and you didn't know. And you'd still have a ladder to beat, one by one if you're good enough to find the correct combination. So persistence and imagination would still be in the mix. But people would actually learn and youtube would stop recommending videos of people doing css battles with only divs and ::after tags because they have no idea how to get started to be good at it.

**Comment by Beowolve** [comment]
This topic needs more attention. I also think it should be allowed to use these CSS features. Just because an attribute uses the SVG syntax doesn't make it SVG. The important aspect here is, that a competition needs to be fair for everyone. And at the moment there are ways to work around the limitations anyway.
If it's not possible to reliably ban a certain feature, then it's better to allow it for fairness reasons alone.

**Reply by chinchang** [reply]
The part "Just because an attribute uses the SVG syntax doesn't make it SVG" is something that interests me. It boils down to that when we say "svg", do we mean the SVG as a tag, or SVG as a syntax/format? We would definitely like to know what others also think about this thought!

> The important aspect here is, that a competition needs to be fair for everyone.

That is true. Issue arises when someone realizes at a later point that SVG was allowed all this while through a not-so-used CSS property, even though the instructions mentioned that SVG isn't allowed. And that most targets could have been easily created with SVG, on which the player spent hours trying to recreate with CSS. At this point it feels like a hack/cheatcode.

**Reply by rafehimself** [reply]
The most important aspect is fairness for everyone

The solution might be triple ladder (it adds unwanted complexity to the page, but automatically solves the current results issue,
cuz otherwise they would need to be removed from the ladder, instead of just being sort out)
CSS only Solution / Unicode Solution / SVG Solution

🤔 In theory it shouldn't be that hard to write the script checking if there is unicode in the solution, that's go to 2nd ladder, if there is SVG path you go to 3th ladder, not sure tho if there is unicode and svg 😄 at the same time, but rather 2 I guess

or simply as @SelenIT suggested if unicode or svg is recognize in the solution, then it only gets max 600 points but nothing more over it,
however that would required a whole ladder rest 🤣

**Reply by gunnarbittersmann** [reply]
> The most important aspect is fairness for everyone

Of course.

The question is: is a feature that a handful people know of fair? Everybody has the same chance to use whatever the HTML and CSS specs provide (and beyond that what the parser allows), right? But ↗︎[equality vs. equity](https://whatwillmatter.com/2017/12/equality-equity/) comes to mind.

> In theory it shouldn't be that hard to write the script checking if there is unicode in the solution

I think this is not true. Just checking the code for non-ASCII characters is certainly not a solution. When `.` (U+002E full stop, ASCII character) is allowed, so should be `·` (U+00B7 middle dot, non-ASCII).

In other words: when `·` would be punished, so should be `.`. Therefore a HTML parser is needed to find all text nodes in the generated DOM. All text nodes need to be checked if any of them contain characters other than whitespace. Whitespace must be allowed since otherwise code formatted for readability would be punished. This raises the question why spaces are allowed to achieve rendered space between two elements when other characters like `_` are not.

This being said, I doubt that a “ladder” system would be feasible.

Even less so for CSS property values taken from SVG. There will always be ways to protect them from being identified as such. Advanced players will find the ways and use them. A ladder system will only increase the gap, I assume.

**Reply by SelenIT** [reply]
IMHO, distinguishing ASCII from non-ASCII characters _could_ make some sense in the past as it made "characters" synonymous to "bytes" in terms of code length (assuming UTF-8 encoding). However,  the ship has sailed since _some_ players used a "hidden way" to submit unicode "as is" and get results measured in characters rather than bytes :)

**Comment by RazvanManolache** [comment]
I think we're beating around the bush a lot on this topic.

The problem when it comes to SVGs is that the guys running the site are not able to do a lot about it. I mean, we're posting text. HTML and CSS are still handled by browsers in a very rudimentary way because of all the dumbass code that exists on the web.

Is HTML an XML? Yes, but hardly, as it has a ton of legacy stuff baked in and all this guesswork about where an unclosed tag ends. Is CSS syntax any good? Hell no, it's just code that's dropped off to secondary algorithms.

Any guy able to read some Javascript can search in the CSS Battle frontend code to see how the validation is done.

 e.match(/svg/gi) || e.match(/clip-path\s*:\s*path/gi) ? {
                           msg: "SVGs are not allowed"

The same is done on the server because.. yeah, how else are they gonna do it? The renderer is not reporting "hey, i did an SVG".

and while you can't write

`<style>*{background:#AC474B;-webkit-box-reflect:left -25em}*>*{background:linear-gradient(#0E1F2B 66q,#FFF 0 9ch,#0E1F2B 0 97q,#FFF 0 139q,#0000 0 45vh,#FFA63F 0 36%,#0000 0 37%,#FFF 0);padding:201 198;clip-path:path('m191 47h21v40h10v5h-13q21 17 8 42 10 3 1 14h5a50 50 2 01-34 89`

you can certainly write 

`<style>*{background:#AC474B;-webkit-box-reflect:left -25em}*>*{background:linear-gradient(#0E1F2B 66q,#FFF 0 9ch,#0E1F2B 0 97q,#FFF 0 139q,#0000 0 45vh,#FFA63F 0 36%,#0000 0 37%,#FFF 0);padding:201 198;clip-path:var(--,path('m191 47h21v40h10v5h-13q21 17 8 42 10 3 1 14h5a50 50 2 01-34 89`

because the var made it not whitespace. 

I know the point of removing SVGs is that a lot of targets had like really good SVG solutions which kinda broke the whole HTML+CSS love. Personally, I like not having SVGs, because they're boring. 

The only problem that I see is the fact that new players have to learn so much to be able to compete. And there's little opportunity to learn. beo was kind enough to make a discord channel that contains a lot of tips and tricks but people still have to find the thing. They still have to understand on which target to use them. Open targets are a start on the site but it's not enough. Letting other people know what tags and properties are used would help create more competition.  I'm not afraid of competition if it makes me better. 

Let's take target 99.

I use a **p** and properties **background**,**border**,**border-radius**,**color**,**font**,**margin**,**padding**,**-webkit-box-reflect** and the unicode character ■. Would that make everybody who knew that be able to get 173 chars? No. Would they know a path they could take? Would they try some stuff and learn something? Hell yeah. Would the information about what tags and properties beo uses make me have 170 chars? No. Would it help me try constructive stuff? Yes. 

We are rarely learning right now and this site could be way more educational if it tried. We have people doing crazy stuff to html+css but nobody knows what. It's knowledge gathered for no purpose. I'm not saying it's amazingly useful what we do, but you know when you first put 0.1+0.2 in the browser console and went "what the hell"? This place has a lot of that but nobody knows. And it's a shame.

**Comment by Beowolve** [comment]
I really see no problem here, I mean, everyone can use what ever feature he thinks works best. If someone doesn't know about box-shadow, or other important features, then it's this guy's issue, right? The same applies to clip-path, and yes, it does share the same syntax with SVG paths, but it's still clip-path. It isn't always the shortest, sometimes polygon beats that even!
The same is true for offset-path and shape-outside. CSS has that features and in a competition where it's all about shortening code, I see no reason to ban useful stuff that does sometimes help to shorten your code.

**Reply by gunnarbittersmann** [reply]
And “to ban” means “to ban *for most, but not for all*.” Some will ways…

The code will be just a little bit longer for some, but for the others this feature will not be available.

I agree with you when you said:
> If it's not possible to reliably ban a certain feature, then it's better to allow it for fairness reasons alone.

**Reply by rafehimself** [reply]
so why is it officially not allowed?
![image](https://user-images.githubusercontent.com/58576132/148054116-2017d43e-4c70-4f89-a634-b38ecc6adbb1.png)

**Reply by gunnarbittersmann** [reply]
And why does it still read “HTML entities” when I use Unicode characters as-is? 🤔
<img width="540" alt="Bildschirmfoto 2022-01-04 um 13 15 19" src="https://user-images.githubusercontent.com/409844/148057814-7c1b515f-3270-4d3d-9f5d-bb678c7833af.png">

**Reply by rafehimself** [reply]
unfortunately "unicode battle" might be more accurate name in a near future

**Comment by SelenIT** [comment]
Maybe it would be worth comparing how much did SVG and/or unicode actually improve the results? Here is some stats I'm aware of (from my memory so might be not precise, doesn't include better results I'm not sure which approach they are based, "CSS" and "Unicode" may include `offset` and `clip-path` other than `path`):

| Target | SVG | CSS | Unicode | Clip-path:path() |
| --- | --- | --- | --- | --- |
| 4 | 105 | 121 | **97** | ? |
| 5 | 142 | 143 | **119** | ? |
| 8 | 163| **157** (was 156) | ? | 170 |
| 10 | 212 | **197** | ? | ? |
| 11 | 158? | 167 | **137** | ~180? |
| 12 | 128 | 158 | **119** | 149? |
| 19 | — | 157 | **128** | 144 |
| 22 | — | 135 | ? | **120** |
| 32 | — | 131 (was 130) | **113** | 151? |
| 41 | — | 155 | **147** | 154 |
| 49 | — | 200 | ? | **175** |
| 54 | — | 275 | ? | **225** |
| 55 | — | **136** | 145 | ~170? |
| 56 | — | 248 | ? | **186** |
| 60 | — | 160? | **99** | 141 |
| 61 | — | 271 | ? | **203** |
| 66 | — | 218 | ? | **160** |
| 70 | — | 242 | **215** | ? |
| 72 | — | 247 | 225 | **206** |
| 80 | — | 183 | **161** | ? |
| 93 | — | ~260 | 224 | **217** |

It appears that all SVG records are broken now, either by "pure CSS" solutions or by Unicode-based solutions. In general, Unicode seems to "win" most of the time, while `clip-path:path()` has maximum advantage for a specific type of targets — symmetrical complex single-color shapes.

Maybe adding a part of a different color to such targets would be a better prevention of overusing `clip-path:path` than filtering the input?..

**Reply by gunnarbittersmann** [reply]
> Maybe adding a part of a different color to such targets would be a better prevention of overusing `clip-path:path` than filtering the input?..

[#85 Pythagoras](https://cssbattle.dev/play/85) comes to my mind.

Fun fact: I had submitted two versions of it for [“Pick My Target”](https://github.com/cssbattle/forum/discussions/98) – one with all squares in the same foreground color and one with different colors for each square. The former can be solved in 121 characters with `clip-path:path()`, while the stripped down version of the current best “pure CSS” solution is 128 characters long.

Of course, they’ve picked the latter, where the `clip-path:path()` solution is much longer (213?).

**Reply by gunnarbittersmann** [reply]
Some additions/updates to this table:
#⁠8: CPP 170
#⁠11: SVG 158? (according to my notes, but they might be wrong)
#⁠19: CPP 144
#⁠32: CSS 131 (was 130), Unicode 113, CPP 151?
#⁠60: CPP 141

**Reply by SelenIT** [reply]
Thanks, updated the table!

**Reply by JDMeijers** [reply]
Update for # 22 : CPP 120

**Reply by SelenIT** [reply]
Updated!

**Comment by rafehimself** [comment]
I've been playing enough of online games in my life to know that without strong admins moderation, cheaters will eventually take over the place and ruin the experience for everybody else ...
Today it's clip-path hack, tomorrow there will be another hack, it's just a matter of time ...

Without integrity and with double standards the whole endeavor doesn't seems like time well spent

**Reply by gunnarbittersmann** [reply]
There. Are. No. Cheaters. Here.

*“The objective of the game is to write HTML/CSS to replicate the given target image in the least code possible.”* To test out the boundaries of what that code might be is part of the game, not cheating.

**Reply by rafehimself** [reply]
what is even more heartbreaking is that it's not some weak random players from the bottom that cheat, but actually guys from top 5 ... which should rather be role models ...

**Reply by gunnarbittersmann** [reply]
Ty zamierzasz grać w code-golfing czy na nerwach?

**Reply by Beowolve** [reply]
You just overestimate the power of clip-path, there is some real optimization going on there, just like on every other part of the code. And it isn't easy to get below CSS on some targets, or even possible. You need to use that right as well.
And stop calling someone a cheater just because they know more tricks than you...

The only reason why there is any discussion here, is because the UI thinks this is a real SVG path and puts a message out there. The checking is very simple implemented and can't distinguish between that. And for me this is a clip-path, not a SVG path, it just uses the SVG syntax here!
You can argue about that understanding and only the site owner can make this clear for all, but for me there is a difference.
The UI also tells u also not to use external images for example and at the same time does prevent you from using embedded images which in my understanding are not external and would be perfectly fine.
There is also a way around that limitation, not that it helps you much...

**Reply by rafehimself** [reply]
https://www.oxfordlearnersdictionaries.com/definition/english/cheat_1
"cheat (at something) to act in a dishonest way in order to gain an advantage, especially in a game, a competition, an exam, etc."

**Reply by gunnarbittersmann** [reply]
Hacking into the database and looking at somebody else’s code or changing scores – that’s acting in a dishonest way.
Where’s the dishonesty in writing code in the editor field and hitting the submit button?

**Reply by rafehimself** [reply]
![image](https://user-images.githubusercontent.com/58576132/148357417-05816445-3f8a-4129-abdc-72faf5bc7e60.png)

you cannot hit submit button unless you cheat to walk around it ... it's pretty straight forward
disrespectful for the admin and other players as well

**Reply by SelenIT** [reply]
However, nothing prevents us from hitting submit button here:
![image](https://user-images.githubusercontent.com/282754/148358665-957aef06-6f2f-416d-88c3-f1a1c35d9553.png)
Same value, same syntax (technically, a CSS string in the SVG `d` attribute format)... just a different property. Which brings us back to the starting question of the topic 🤷

**Reply by gunnarbittersmann** [reply]
> you cannot hit submit button unless you cheat to walk around it

You can hit the submit button when using a CSS feature.

The more you know about CSS, the higher you score. That’s what the game is all about, isn’t it?

If using a CSS feature is cheating to you, what is 
- using `10q` for `10px`? It’s not the same value, but might work. (Even `9q` might do.)
- tweaking color values, eg. `#A64940` for `#A64942` in order to save one character in `<body text=A64A4>`?
- using RGBA colors?

Should all the solutions taking advantage of slightly off values be invalid? After mourning about false negatives you would mourn about false positives?

What you call “cheating” I call “code-golfing.”

**Comment by chinchang** [comment]
Hey folks! We really appreciate the thoughts everyone is putting in here. That is the reason why CSSBattle continuously improves. Thank you for that! But we just wanted to request everyone to be calm and not get personal or get affected personally. 
The true reason why we are able to improve the rules is that we ourselves find some new way which some players used and then after debating we consider it not appropriate. So it's essentially you folks and your curiosity n eagerness to level up that makes the game better.

The point under discussion is a really important topic. Needless to say, we need inputs from everyone to come up with the best possible decision here. It isn't easy to simply change the rule, because they are mostly irreversible. At the same time, we want to get as fair as possible. Please keep the suggestions/feedback coming. We are actively going through them all. 🙏🏽

**Comment by rafehimself** [comment]
so why nothing has been done ? this is clearly svg cheat?

`<style>*{background:#000}* *{background:#D4392D;clip-path:var(--,path('m127 94a50 60-30 115,90q20-25,4-45,18-33-8-45h59V62h10v36a49 60,30 11-9,9`

bunch of numbers supposed to represent css skills? 🤡

**Reply by RazvanManolache** [reply]
Well, it is a good question. Maybe we should ban path altogether. I honestly see no way of them preventing it without a ton of effort. 

At least I'm happy it's now on an open target for everyone to know about it. While it's a cheat to the principle of the game, it's more of a cheat if nobody knows about it.

**Reply by gunnarbittersmann** [reply]
> > so why nothing has been done ?
>
> Well, it is a good question.

No, it is not.

The makers of the game have done everything they could to prevent from using `clip-path: path()`. Obviously, that’s not enough; some players found a way…

> Maybe we should ban path altogether.

How?

> I honestly see no way of them preventing it without a ton of effort.

Now matter what they would try to implement, some players will still find a way…

> While it's a cheat to the principle of the game, it's more of a cheat if nobody knows about it.

I’m not sure about the first part. It’s part of the CSS spec. Is it a cheat then?

But I agree with you on the second part.

The only way to justice that I see is to take away the restriction, so that the feature becomes available for everybody without using hacks like ‘hiding’ the path in a custom property that only a few might be aware of.

**Comment by SelenIT** [comment]
It occured to me that a possible way to distinguish "SVG-in-CSS" solutions from "non-SVG" could be detecting the _path string itself._ It's always quoted (at least at the beginning), always starts with the "M" letter followed by two numbers, optionally separated by spaces, doesn't support escaping, and seems to be the only context where such sequence may appear in CSS. So it can be detected with a relatively simple RegExp pattern (like `/['"]\s*m\s*[\d.]+\s*[\d.]+/i`), regardless the way it's assigned to the CSS property (directly or through `var()`).

A side effect of this approach will be that solutions using `offset:path()` will be considered "SVG-containing" as well. But I assume it would be consistent and fair after all. Especially if both will just have special scoring rules/table (as I have been hoping for a long time) rather than be banned totally.

**Reply by chinchang** [reply]
That definitely sounds like a much better idea than checking for properties!
